p2p_beef_tx.go

```diff
@@ -0,0 +1,229 @@
+package paymail
+
+import (
+	"encoding/hex"
+	"errors"
+
+	"github.com/libsv/go-bt/v2"
+)
+
+type CompoundMerklePath []map[string]uint64
+
+type CMPSlice []CompoundMerklePath
+
+const (
+	BEEFMarkerPart1 = 0xBE
+	BEEFMarkerPart2 = 0xEF
+)
+
+const (
+	HasNoCMP = 0x00
+	HasCMP   = 0x01
+)
+
+type TxData struct {
+	Transaction *bt.Tx
+	PathIndex   *bt.VarInt
+}
+
+type DecodedBEEF struct {
+	CMPSlice        CMPSlice
+	InputsTxData    []TxData
+	ProcessedTxData TxData
+}
+
+func DecodeBEEF(beefHex string) (*DecodedBEEF, error) {
+	beefBytes, err := extractBytesWithoutVersionAndMarker(beefHex)
+	if err != nil {
+		return nil, err
+	}
+
+	cmpSlice, remainingBytes, err := DecodeCMPSliceFromStream(beefBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	transactions, err := DecodeTransactionsWithPathIndexes(remainingBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(transactions) == 0 {
+		return nil, errors.New("no transactions found")
+	}
+
+	// get the last transaction as the processed transaction - it should be the last one because of khan's ordering
+	processedTx := transactions[len(transactions)-1]
+
+	transactions = transactions[:len(transactions)-1]
+
+	return &DecodedBEEF{
+		CMPSlice:        cmpSlice,
+		InputsTxData:    transactions,
+		ProcessedTxData: processedTx,
+	}, nil
+}
+
+func DecodeCMPSliceFromStream(hexBytes []byte) (CMPSlice, []byte, error) {
+	if len(hexBytes) == 0 {
+		return nil, nil, errors.New("provided hexStream is empty")
+	}
+
+	nCMPs, bytesUsed := bt.NewVarIntFromBytes(hexBytes)
+	hexBytes = hexBytes[bytesUsed:]
+
+	var cmpPaths []CompoundMerklePath
+	for i := 0; i < int(nCMPs); i++ {
+		cmp, bytesUsedToDecodeCMP, err := NewCMPFromStream(hexBytes)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		cmpPaths = append(cmpPaths, cmp)
+		hexBytes = hexBytes[bytesUsedToDecodeCMP:]
+	}
+
+	cmpSlice := CMPSlice(cmpPaths)
+
+	return cmpSlice, hexBytes, nil
+}
+
+func NewCMPFromStream(hexBytes []byte) (CompoundMerklePath, int, error) {
+	height, bytesUsed, err := extractHeight(hexBytes)
+	if err != nil {
+		return nil, 0, err
+	}
+	hexBytes = hexBytes[bytesUsed:]
+
+	var cmp CompoundMerklePath
+	previousHeight := height
```

Comment:
Something like a `currentHeight` or `runningHeight` sounds better imo in this case.

Reply:
done - it's currentHeight now